File: Algorithms/some_intern_contests/tinkoff_winter/task_7/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

const int MOD = 998244353;

std::vector<std::vector<int>> build_binomial(int max_n) {
    std::vector<std::vector<int>> comb(max_n + 1, std::vector<int>(max_n + 1, 0));
    for (int i = 0; i <= max_n; ++i) {
        comb[i][0] = 1;
        for (int j = 1; j <= i; ++j) {
            comb[i][j] = (comb[i-1][j-1] + comb[i-1][j]) % MOD;
        }
    }
    return comb;
}

std::vector<int> optimized_solve(int n, int k, const std::vector<int>& a) {
    std::vector<long long> S(2*k + 1, 0);
    for (int x : a) {
        long long cur = 1;
        for (int l = 0; l <= 2*k; ++l) {
            S[l] = (S[l] + cur) % MOD;
            cur = (cur * x) % MOD;
        }
    }

    auto comb = build_binomial(2*k);
    
    int inv2 = (MOD + 1) / 2;
    
    std::vector<int> answers(k + 1, 0);
    for (int p = 1; p <= k; ++p) {
        long long res = 0;
        if (p % 2 == 0) {
            int q = p / 2;
            for (int r = 0; r < q; ++r) {
                long long tmp = ((S[r] * S[p-r]) % MOD - S[p] + MOD) % MOD;
                tmp = (tmp * comb[p][r]) % MOD;
                res = (res + tmp) % MOD;
            }
            long long tmp = ((S[q] * S[q]) % MOD - S[p] + MOD) % MOD;
            tmp = (tmp * inv2) % MOD;
            tmp = (tmp * comb[p][q]) % MOD;
            res = (res + tmp) % MOD;
        } else {
            int q = (p - 1) / 2;
            for (int r = 0; r <= q; ++r) {
                long long tmp = ((S[r] * S[p-r]) % MOD - S[p] + MOD) % MOD;
                tmp = (tmp * comb[p][r]) % MOD;
                res = (res + tmp) % MOD;
            }
        }
        answers[p] = res;
    }
    
    return std::vector<int>(answers.begin() + 1, answers.end());
}

int main() {
    int n, k;
    std::cin >> n >> k;
    std::vector<int> a(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> a[i];
    }
    auto answers = optimized_solve(n, k, a);
    for (int answer : answers) {
        std::cout << answer << " ";
    }
    std::cout << std::endl;
    return 0;
}

File: Algorithms/some_intern_contests/tinkoff_winter/task_7/solution.py
def build_binomial(max_n):
    """
    Build a 2D table comb so that comb[n][r] = C(n, r) mod MOD
    for all n, r up to max_n.
    """
    comb = [[0]*(max_n+1) for _ in range(max_n+1)]
    for i in range(max_n+1):
        comb[i][0] = 1
        for j in range(1, i+1):
            comb[i][j] = (comb[i-1][j-1] + comb[i-1][j]) % 998244353
    return comb

def solve():
    n, k = map(int, input().split())
    a = list(map(int, input().split()))
    MOD = 998244353
    
    # Precompute power sums S[l] = sum of b[i]^l for l=0..2k
    # This is potentially O(m * 2k), which can be large but is usually still feasible.
    S = [0]*(2*k+1)
    for x in a:
        cur = 1
        for l in range(2*k+1):
            S[l] = (S[l] + cur) % MOD
            cur = (cur * x) % MOD

    # Precompute binomial coefficients up to 2k
    comb = build_binomial(2*k)
    
    inv2 = (MOD + 1)//2   # modular inverse of 2 mod 998244353
    
    # Compute answers for p = 1..k
    answers = [0]*(k+1)
    for p in range(1, k+1):
        # Use the formulas described:
        # If p = 2q (even):
        #     f(p) = sum_{r=0..q-1} [ C(p,r) * (S[r]*S[p-r] - S[p]) ]
        #            + C(p,q) * ((S[q]^2 - S[p])/2)
        # If p = 2q+1 (odd):
        #     f(p) = sum_{r=0..q} [ C(p,r) * (S[r]*S[p-r] - S[p]) ]
        
        res = 0
        if p % 2 == 0:
            # even p = 2q
            q = p // 2
            for r in range(q):
                tmp = ((S[r] * S[p-r]) % MOD - S[p]) % MOD
                tmp = (tmp * comb[p][r]) % MOD
                res = (res + tmp) % MOD
            # middle term r = q
            tmp = ((S[q]*S[q]) % MOD - S[p]) % MOD
            tmp = (tmp * inv2) % MOD       # divide by 2
            tmp = (tmp * comb[p][q]) % MOD
            res = (res + tmp) % MOD
        else:
            # odd p = 2q+1
            q = (p - 1)//2
            for r in range(q+1):
                tmp = ((S[r] * S[p-r]) % MOD - S[p]) % MOD
                tmp = (tmp * comb[p][r]) % MOD
                res = (res + tmp) % MOD
        
        answers[p] = res % MOD
    
    # Output the answers
    out = []
    for p in range(1, k+1):
        out.append(str(answers[p]))
    print("\n".join(out))


if __name__ == "__main__":
    solve()